### PR TITLE
feat(crap4ts): #182 — OxcWalker minimal cyclomatic walker (6 decision points) + parse-failure pattern

### LINKED ISSUE #182: feat(crap4ts): OxcWalker minimal cyclomatic walker (6 decision points) + parse-failure pattern
## Summary

Replace `OxcWalker::extract`'s `unimplemented!()` body at `crates/crap4ts/src/adapters/walker/mod.rs` with a real oxc-based AST visitor that emits `Vec<FunctionComplexity>` for `.ts` source. Handle the 6 universal cyclomatic decision points (`if`, `for`/`for-of`/`for-in`, `while`, `do-while`, `switch case`, `&&`/`||`). TS-specific points (`?:`, `?.`, `??`, `catch`, JSX) deferred to W2.1. Parse failures emit `Err(CrapError::SourceParse(file_path: msg))` per the crap4rs syn-walker pattern — orchestrator at `crap-core/src/core/mod.rs:286-310` handles continue-and-diagnose.

## Acceptance Criteria

- [ ] `cargo nextest run --workspace --all-targets -E 'test(walker_smoke)'` passes — all 6+ decision-point tests + nested-function test + parse-failure test green
- [ ] `cargo nextest run --workspace --all-targets` passes — no regressions
- [ ] `cargo +1.93 check --workspace --all-targets` + `cargo fmt --check` + `cargo clippy --workspace --all-targets -- -D warnings` + `cargo doc --workspace --no-deps -D warnings` + `cargo deny check` all pass
- [ ] AST-purity grep clean on `crates/crap-core/src/`
- [ ] Wire snapshot byte-identical (no envelope shape change in W1.2)
- [ ] `#[should_panic]` `oxc_walker_extract_unimplemented` test removed
- [ ] PR body lists the exact oxc version pinned + any API name shifts from breadboard-referenced names (via Context7 pre-flight)
- [ ] Hard circuit breaker enforced: NO `ConditionalExpression`, `OptionalMemberExpression`, `ChainExpression`, `LogicalExpression(Coalesce)`, `TryStatement`/`CatchClause`, `JSXExpressionContainer` arms counted — those land in W2.1 only

## Context

- Epic: #173
- Pipeline: `crap-rs/20260513-typescript-adapter`
- Plan: `~/Github/ops/workspace/crap-rs/20260513-typescript-adapter/impl-plan.md` — see section **W1.2** for full scope, decision-point mapping, oxc API pre-flight, and worktree prompt
- Sibling: W1.1 (Istanbul parser; either can open first)
- Related ADRs: D1 (oxc choice), D3 (port purity), D8 (parse-failure pattern)

## Discovery

- Confirm `oxc_parser::Parser::new(alloc, source, source_type).parse()` API + `oxc::ast::Statement`/`Expression` enum variant names match breadboard references via Context7 lookup as first step in the worktree
- Verify `SourceType::ts()` constructor (vs. `SourceType::from_path` canonical dispatcher) — CEng ADVISORY-3 prefers `SourceType::from_path` if exposed; verify at implementation time

### PR DESCRIPTION
**Pull request by cmbays**
## Summary

Replace `OxcWalker.extract`'s `unimplemented!()` body with a real oxc-based AST visitor that emits one `FunctionComplexity` per function declaration / function expression / arrow function / class method. The walker counts the six universal cyclomatic decision points (`if`, `for` / `for-of` / `for-in`, `while`, `do-while`, `switch case`, `&&` / `||`) into existing `ContributorKind` variants — no new enum variants land. TS-specific decision points (`?:`, `?.`, `??`, `catch`, JSX) are intentionally deferred to W2.1 / ADR (a) per the hard circuit breaker; the walker traverses past them without emitting contributors. Parse failures bubble up as `Err(CrapError::SourceParse("file_path: <oxc msg>"))` so the orchestrator at `crap-core/src/core/mod.rs:286-310` can continue past a malformed file and increment `AnalysisDiagnostics.files_unparseable`.

Self-CRAP on `crates/crap4ts/src/` worst-CC dropped from 44 (monolithic `visit_statement`) to 9 (`visit_for_init`) by splitting per-construct helpers in the syn-walker pattern — file weight is now comparable to `crap4rs::adapters::complexity` (`parse_brda` CC=11 is the workspace baseline). 12 `walker_smoke.rs` tests ground-truth every BDD-feature contract for W1.2 scope at the unit level (cucumber harness lands in W3.3).

## Acceptance gates

- [x] `cargo build --workspace`
- [x] `cargo nextest run --workspace --all-targets -E 'binary(walker_smoke)'` — 12 tests green
- [x] `cargo nextest run --workspace --all-targets` — 983 / 983 (was 972 / 972 pre-W1.2; +11 net = +12 new walker_smoke − 1 removed `#[should_panic]`)
- [x] `cargo +1.93 check --workspace --all-targets`
- [x] `cargo fmt --check && cargo clippy --workspace --all-targets -- -D warnings`
- [x] `RUSTDOCFLAGS=-D warnings cargo doc --workspace --no-deps`
- [x] `cargo deny check`
- [x] AST-purity on `crap-core/src/` clean (no `oxc` / `syn` / `tree_sitter` / `swc` imports)
- [x] Wire snapshot byte-identical (`wire_envelope_snapshot::envelope` PASS)
- [x] `#[should_panic]` `oxc_walker_extract_unimplemented` test removed
- [x] oxc pinned at workspace `0.129` (no breaking API shifts; one secondary-source correction surfaced — see below)

## oxc API verification (Context7)

Verified against `oxc-project/oxc` (Context7 library id `/oxc-project/oxc`, oxc 0.129 source in cargo registry). The umbrella `oxc` crate re-exports `oxc_allocator`, `oxc_ast`, `oxc_parser`, `oxc_span`, `oxc_syntax`; the trait `Visit` lives behind a feature flag (`oxc/ast_visit`), which is NOT enabled in our workspace. We use manual recursion per the plan-of-record, matching the syn-walker shape.

**Confirmed API surface (no drift from training data):**
- `oxc::allocator::Allocator::default()`
- `oxc::parser::Parser::new(&alloc, source, source_type).parse() -> ParserReturn` with `ParserReturn.errors: Vec<OxcDiagnostic>` and `OxcDiagnostic: Display` writing the message
- `oxc::span::SourceType::from_path(P: AsRef<Path>) -> Result<Self, UnknownExtension>` (used; CEng ADVISORY-3 was correct to predict this — W2.2's dispatch becomes a near-no-op)
- `oxc::span::SourceType::ts()` (fallback for unrecognised extensions)
- `Statement::{If, For, ForIn, ForOf, While, DoWhile, Switch}Statement` — variant names match the plan
- `Expression::LogicalExpression` with `oxc::syntax::operator::LogicalOperator::{And, Or, Coalesce}` — matches the plan
- `SwitchCase { test: Option<Expression>, consequent: Vec<Statement> }` — `default:` arm has `test: None` (NOT scored, per the BDD outline row semantics)

**Drift correction (surfaced during implementation):** the W1.2 prompt and impl-plan §W1.2 both claim oxc `Span { start, end }` is inclusive on both ends and that the walker should add `+1` to `end_line` at the domain boundary. Primary sources contradict:
- `oxc_span/src/span.rs` doc-comment: `start` is the zero-based start; `end` is the zero-based end position (exclusive — empty span has `start == end`).
- `oxc_span/README.md` (Context7-fetched): "spans defined as inclusive of the start position and exclusive of the end position".
- `crap-core::domain::types::SourceSpan` (line 8): `end_line` is 1-based **inclusive**.
- `crap4rs::adapters::complexity::span_of` uses `sp.end().line` directly with NO `+1`.

Correct conversion is therefore `end_line = byte_to_line(source, span.end.saturating_sub(1))` — the 1-based line of the LAST BYTE in the span. The walker uses this convention. Recorded in `ops/workspace/crap-rs/20260513-typescript-adapter/observations.md` as a plan-of-record pin.

## Test plan

`crates/crap4ts/tests/walker_smoke.rs` (12 tests, all `binary(walker_smoke)`-matchable):

- **Baseline** (`simple.ts` — W1.1 fixture reused per CQO ADVISORY-8):
  - `simple_function_scores_cyclomatic_one_with_no_contributors` — CC=1, zero contributors, metric label preserved
- **Six universal decision points** (one fixture per construct family):
  - `if_branch_contributes_one_decision_point` — `IfBranch` at expected line
  - `classic_for_statement_contributes_one_for_loop` / `for_of_statement_contributes_one_for_loop` / `for_in_statement_contributes_one_for_loop` — all three for-flavours collapse to `ForLoop`
  - `while_statement_contributes_one_while_loop` / `do_while_statement_contributes_one_do_while_loop`
  - `switch_case_contributes_one_case_branch_excluding_default` — `case 1:` counts, `default:` does NOT
  - `logical_and_contributes_one_logical_operator` / `logical_or_contributes_one_logical_operator` — `&&` and `||` only (NOT `??`)
- **Function discovery**:
  - `nested_functions_are_separate_complexity_sites` — `outer`'s `if` does NOT leak into `inner`'s score, and vice versa
- **Parse-failure path**:
  - `malformed_typescript_returns_source_parse_with_file_prefix` — `Err(CrapError::SourceParse("broken.ts: <oxc msg>"))`, orchestrator-compatible

7 new `.ts` fixtures land under `crates/crap4ts/tests/fixtures/ts-fixtures/`; `simple.ts` is reused from W1.1.

## Wire-shape canary

**Byte-identical.** Adapter-only changes; no domain or `crap-core` source touched. `wire_envelope_snapshot::envelope` re-passes against the pre-W1.1-merge snapshot.

## Plan deviations (W1.1 discipline — surface, don't silently reinterpret)

1. **Filter expression in acceptance gates:** the W1.2 prompt and impl-plan §W1.2 both spec `cargo nextest run --workspace --all-targets -E 'test(walker_smoke)'`. Empirically, nextest's `test()` filter matches by test-name substring (so it matched zero tests because no individual `#[test]` is named `walker_smoke_*`). The correct filter is `binary(walker_smoke)`, which matches integration-test binaries. Same class of plan-of-record drift as W1.1's "aggregated-fixture" delegation prompt — flagging here per the rule established in W1.1's PR.
2. **Span semantics +1 correction** — see "oxc API verification" above.
3. **Self-CRAP refactor surfaced during implementation:** the first pass of `visit_statement` was a monolithic CC=44 cognitive match. Refactored into per-construct helpers (`visit_for`, `visit_while`, `visit_do_while`, `visit_switch`, `visit_try`, plus expression-side `visit_conditional` / `visit_arguments` / `visit_array_elements` / `visit_each`) before commit; worst walker function is now CC=9 (`visit_for_init`), comparable to the syn walker's CC=11 (`parse_brda`). Not a plan deviation per se — the plan said "manual recursion" without specifying decomposition depth — but worth noting as a self-CRAP win.

Closes #182

## Summary by CodeRabbit

* **New Features**
  * Implemented a complete TypeScript complexity analysis engine that measures cyclomatic complexity across control flow structures.
  * Added support for analyzing functions, loops, conditionals, switches, and logical operators.
  * Implemented error handling for malformed code with detailed parse error reporting.

* **Tests**
  * Added comprehensive test fixtures and smoke tests covering various code patterns and complexity scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/breezy-bays-labs/crap-rs/pull/198)

### CODE REVIEW

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 1]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+/// label positions). `byte_to_line` is reused for the line count so
+/// the two helpers stay self-consistent on EOF clamp behaviour.
+fn byte_to_line_col(source: &str, byte: u32) -> (usize, u32) {
+    let limit = (byte as usize).min(source.len());
+    let line = byte_to_line(source, byte);
+    let prefix = &source[..limit];
+    let line_start = prefix.rfind('\n').map(|i| i + 1).unwrap_or(0);
+    let column = (limit - line_start) as u32 + 1;
+    (line, column)
+}
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:900]: Addressed in commit 06b16ff — switched `byte_to_line_col` to a byte-level scan via `source.as_bytes()` + `rposition(b'\n')`, mirroring `byte_to_line`'s panic-safe pattern. Column semantics ("UTF-8 code units / bytes") preserved. Added `unicode_identifiers_do_not_panic_in_span_to_column_conversion` regression test using `function π(αβγ)` — would have panicked on the prior `&source[..limit]` implementation. Thanks for the catch — real panic-safety bug.

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 2]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+            .id
+            .as_ref()
+            .map(|bi| bi.name.as_str().to_string())
+            .or(name_hint)
+            .unwrap_or_else(|| "<anonymous class>".to_string());
+        for element in &class.body.body {
+            if let ClassElement::MethodDefinition(method) = element {
+                self.record_method(method, &class_name);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:239]: Real soundness gap on modern TS (class-field auto-binding arrow handlers, static initialization blocks). Declined for this PR to keep scope on the 6 universal cyclomatic decision points — filed as **#199** (`type:bug priority:soon`, blocked-by #182) so it lands before or alongside W2.1 (#184) where the test surface naturally exercises class-field arrows.

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 3]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+                }
+            }
+
+            // Everything else is structural-only at this scope: leaf
+            // statements (break/continue/debugger/empty/with), bare
+            // module statements (re-exports / imports / TS-flavoured
+            // assignment-exports), and TypeScript type-only
+            // declarations all contribute no executable decision
+            // points to the enclosing function body.
+            _ => {}
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:352]: Split disposition:

- **`TSModuleDeclaration`** — real gap (TS namespaces can host functions). Filed as part of the traversal-gaps umbrella **#200** (`type:polish priority:later`).
- **`WithStatement`** — declining. `with` is forbidden in ES modules and in any strict-mode source; TypeScript modules are implicitly strict and the TS compiler emits an error on `with` in module sources. Real-world `.ts` files won't reach this AST node through a successful parse; not worth a code path.

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 4]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+            }
+            Expression::SequenceExpression(seq) => self.visit_each(&seq.expressions, out, nesting),
+            Expression::UnaryExpression(u) => self.visit_expression(&u.argument, out, nesting),
+            Expression::BinaryExpression(b) => {
+                self.visit_expression(&b.left, out, nesting);
+                self.visit_expression(&b.right, out, nesting);
+            }
+            Expression::AssignmentExpression(a) => {
+                self.visit_expression(&a.right, out, nesting);
+            }
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:650]: Edge case (`x[foo()] = 1` where the index expression has nested functions). Filed under the traversal-gaps umbrella **#200** (`type:polish priority:later`).

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 5]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+            Expression::StaticMemberExpression(m) => {
+                self.visit_expression(&m.object, out, nesting);
+            }
+            Expression::PrivateFieldExpression(m) => {
+                self.visit_expression(&m.object, out, nesting);
+            }
+
+            // Leaves + JSX (out of scope for W1.2) — no scoring, no
+            // recursion. JSX bodies get full coverage in W2.1.
+            _ => {}
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:697]: Same edge-case category as the `AssignmentExpression` LHS finding. Filed under the traversal-gaps umbrella **#200** (`type:polish priority:later`).

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 6]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+        &mut self,
+        obj: &ObjectExpression<'_>,
+        out: &mut Contributors,
+        nesting: Option<u32>,
+    ) {
+        for prop in &obj.properties {
+            match prop {
+                ObjectPropertyKind::ObjectProperty(p) => {
+                    self.visit_expression(&p.value, out, nesting);
+                }
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:824]: CodeRabbit raised the same gap with the explicit "feel free to defer" recommendation. Filed under the traversal-gaps umbrella **#200** (`type:polish priority:later`).

File: crates/crap4ts/src/adapters/walker/mod.rs [hunk 7]
```diff
@@ -34,14 +98,863 @@ impl Default for OxcWalker {
...
+/// `span.end.saturating_sub(1)` on a zero-length span at offset 0)
+/// never panic.
+fn byte_to_line(source: &str, byte: u32) -> usize {
+    let limit = (byte as usize).min(source.len());
+    source.as_bytes()[..limit]
+        .iter()
+        .filter(|&&b| b == b'\n')
+        .count()
+        + 1
+}
```
[the remark this answers is not in the document]
Comment (cmbays) [reply crates/crap4ts/src/adapters/walker/mod.rs:882]: Declining for this PR:

1. **The cited rule is about I/O streaming, not in-memory scan complexity.** The reference to "prefer streaming interfaces or passing file paths" is talking about avoiding `fs::read_to_string` on large files at I/O time — it doesn't apply to in-memory `&str` scans after the file is already loaded. The walker port shape is `extract(source: &str, file_path: &str, metric)` per ADR D9; that's a (separately tracked) port-evolution issue (F7 / #179).
2. **Walking-skeleton stage.** Pre-mature optimization without a measurement. Real cost will be visible on the crap4ts@1.x corpus parity oracle (W3.2 / #190) — if the harness shows measurable walker latency on the ~3000-LOC corpus, we revisit with a real benchmark + a line-index helper. If it doesn't, the O(M·N) walk is fine for the file sizes crap4ts targets.

Filing under `type:perf` will happen if + when W3.2 surfaces it.
